### PR TITLE
fix(linux): Remove `ibus-keyman.post{inst,rm}` and verify that `ibus-daemon` is running when we start km-config

### LINKED ISSUE #8996: bug(linux): How useful is debian/ibus-keyman.postinst?
From [Debian bug #1034040](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1034040):

> Hi!
> 
> Without being sure I actually found a bug, I would like to raise a 
> discussion as a follow-up of <https://bugs.debian.org/1033571> (which 
> hasn't been well received so far). The unblock discussion made me aware 
> of this script:
> 
> https://github.com/keymanapp/keyman/blob/master/linux/debian/ibus-keyman.postinst
> 
> The script makes me think of im-config, which sets the applicable 
> environment variables and starts ibus-daemon at login.
> 
> I suppose the purpose with the script is to make ibus-keyman instantly 
> usable in the same session as it is installed. I imagine that may work 
> if ibus-daemon was already running and thus the needed environment 
> variables already set for the session. But what if ibus was pulled as a 
> dependency when installing ibus-keyman? Will keyman work in a meaningful 
> way even if the variables are not present?
> 
> If the answer to that question is "no", wouldn't it be more 
> straightforward to instruct the users to relogin before starting to use 
> keyman, and thus drop the postinst script?
> 
> I may well have misunderstood it, and if so, I would appreciate a brief 
> explanation of the intention with the script.
> 
> -- 
> Rgds,
> Gunnar

### PR DESCRIPTION
**Pull request by ermshiperete**
The reason we had `ibus-keyman.postinst` previously was so that ibus (and thus ibus-engine-keyman) is running after the installation of Keyman, so that the user can start to use Keyman immediately. However, at the time the .postinst script is called it is not guaranteed that all dependencies are completely set up. In fact, that's what causes #10478, although we could work around that. 

An alternative approach that this PR implements is that we check that ibus is running when the user tries to install a keyboard. This has the additional benefit that this check happens every time `km-config` or `km-package-install` get called, so it also helps in case ibus isn't running at a later point in time.

Fixes #8996 and #10478.

# User Interface Changes

If ibus is not running (and `fcitx` is not running) when `km-config` gets started, a message box is displayed:

Screenshot from 2024-02-26 17-10-55

If the user chooses "No" `km-config` exits. If the user selects "Yes" we try to start ibus and if it's running now show the `km-config` UI. Otherwise we display another message box telling the user he has to reboot or logout and then exit `km-config`.

With `km-package-install` if ibus is not running (and `fcitx` is not running either) the program exits with the following error:

```
ERROR:km-package-install: error: ibus-daemon is not running. You might have to reboot or logout and login again.
```

# User Testing

Please restore to a snapshot with the state immediately after system installation before running each test!

- Download and unzip the artifacts
- For each test, install Keyman with these commands in the directory that contains the artifacts:

  ```bash
  sudo apt update
  sudo apt install ./*$(lsb_release -c -s)*.deb
  ```

- You should **not** reboot after installing Keyman or keyboards

## Tests

The tests should be run on:

  - **GROUP_FOCAL**: Ubuntu 20.04 Focal with Gnome Shell and X11
  - **GROUP_NOBLE_X11**: Ubuntu 24.04 Noble with Gnome Shell and X11
  - **GROUP_NOBLE_WAYLAND**: Ubuntu 24.04 Noble with Gnome Shell and Wayland

**TEST_REG_INST_UI**: Regular installation works and no message appears in `km-config`

- install Keyman with the above command
- start km-config
- verify that no message about non-running ibus-daemon shows up
- install a Keyman keyboard
- verify that you can switch to the keyboard and the keyboard works

**TEST_REG_INST_CLI**: Regular installation works and no message appears with `km-package-install`

- install Keyman with the above command
- install a keyboard with `km-package-install`:
  ```bash
  km-package-install -p khmer_angkor --bcp47 km
  ``` 
- verify that no message about non-running ibus-daemon shows up (ignore the DeprecationWarning that shows up, that is reported in #10850)
- verify that you can switch to the keyboard and the keyboard works

**TEST_NOIBUS_INST_UI**: Installation works if ibus is not running and a message appears in `km-config`

- stop ibus by running `ibus exit`
- install Keyman with the above command
- start `km-config`
- verify that a message displays that ibus-daemon is not running and if you want to start ibus-daemon now.
- press "No". Verify that `km-config` exits
- start `km-config` again
- verify that a message displays that ibus-daemon is not running and if you want to start ibus-daemon now.
- this time press "Yes"
- install a Keyman keyboard
- verify that you can switch to the keyboard and the keyboard works

**TEST_NOIBUS_INST_CLI**:  Installation works if ibus is not running and a message appears with `km-package-install`

- stop ibus by running `ibus exit`
- install Keyman with the above command
- install a keyboard with `km-package-install`:
  ```bash
  km-package-install -p khmer_angkor --bcp47 km
  ``` 
- verify that the following error is shown and nothing else happens:
  ```
  ERROR:km-package-install: error: ibus-daemon is not running. You might have to reboot or logout and login again.
  ```
  (ignore the DeprecationWarning that shows up, that is reported in #10850)
- start ibus with `ibus start`
- run `km-package-install` again:
  ```bash
  km-package-install -p khmer_angkor --bcp47 km
  ``` 
- verify that you can switch to the keyboard and the keyboard works

### CODE REVIEW

**Review by mcdurdin — APPROVED** [review]

File: linux/keyman-config/keyman_config/ibus_util.py [hunk 1]
```diff
@@ -11,11 +12,22 @@
...
 
 
+class IbusDaemon(Enum):
+    """
+    States of the ibus-daemon
+    """
+    RUNNING = 0,        'ibus-daemon is running'
+    NOT_RUNNING = 1,    'ibus-daemon is not running'
+    MORE_THAN_ONE = 2,  'Error: more than one instance of ibus-daemon is running'
+    ERROR = 3,          'Error: invalid/unknown user or exception'
```
Comment (darcywong00) [inline linux/keyman-config/keyman_config/ibus_util.py:28]: Do you want a separate error for unknown user / fail to SUDO?

Comment (ermshiperete) [reply linux/keyman-config/keyman_config/ibus_util.py]: Done.

**Comment by bharanidharanj** [comment]
# Test Results

### GROUP_FOCAL: Ubuntu 20.04 Focal with Gnome Shell and X11

* **TEST_REG_INST_UI (PASSED):** Tested with the attached PR build 17.0.279-beta (package version 17.0.279-1~PR-10788-2472.1+focal1) on Ubuntu 20.04 Focal Linux OS (VM) and here is my observation: 1. Installed the Keyman with the above mentioned command. 2. Run km-config command on the terminal window and verified that there was no message about non-running ibus-daemon showing up. 3. Installed a (Tamil 99 Basic) keyman keyboard. 4. Verified that the switching between keyboards works correctly. 

   


* **TEST_REG_INST_CLI (PASSED):** 1. Run the 'km-package-install' for khmer_angkor keyboard command on the terminal window. 2. Verified that there was no message about non-running ibus-daemon showing up. 3. Verified that the switching between keyboards works correctly.

**Comment by bharanidharanj** [comment]
# Test Results

### GROUP_FOCAL: Ubuntu 20.04 Focal with Gnome Shell and X11

* **TEST_NOIBUS_INST_UI (PASSED):** 1. Stopped ibus by executing 'ibus exit' command. 2. Run 'km-config' command on the terminal. 3. Verified that a message displayed that "ibus-daemon is not running.." on the screen. 4. Pressed the "No" button. Verified that the km-config window closed. 5. Opened Keyman configuration dialog by running km-config command on the terminal. 6. Verified that a message displayed "ibus-daemon is not running.." on the screen. 7. Clicked the 'Yes' button. 8. Installed a Keyman (sil_ipa) keyboard. Verified that we can switch to sil_ipa keyboard and the keyboard is working as expected.

**Comment by bharanidharanj** [comment]
# Test Results

### GROUP_FOCAL: Ubuntu 20.04 Focal with Gnome Shell and X11

* **TEST_NOIBUS_INST_CLI (FAILED):** 1. Stopped ibus by running ibus exit command. 2. Installed Khmer_Angkor keyboard by running the above command. 3. Verified that it is showing an error message on the terminal. 4. Tried to start ibus by running 'ibus start' command and noticed that it is showing some warning message on the screen. Seems to be an issue. 

..error message on the terminal

   


..ibus start warning message

**Comment by bharanidharanj** [comment]
# Test Results

### GROUP_NOBLE_X11: Ubuntu 24.04 Noble with Gnome Shell and X11

* **TEST_REG_INST_UI (PASSED):** Tested with the attached PR build (17.0.279-beta (package version 17.0.279-1~PR-10788-2472.1+noble1) on Ubuntu 24.04 Noble OS (X11) (Virtual Box) and here is my observation: 1. Installed a Keyman keyboard using km-config command. 2. Verified that there are no messages about non-running ibus-daemon shows up. 3. Verified that it is possible to switch keyboard and the keyboard is working fine. 

   

* **TEST_REG_INST_CLI (PASSED):** 1. Opened the terminal window. 2. Installed Khmer Angkor keyboard using the km-package-install command on the terminal. Verified that no message about non-running ibus-daemon shows up on the terminal window. 3. Also, verified that we can switch to the keyboard and the keyboard is working fine. 

   

* **TEST_NOIBUS_INST_UI (PASSED):** 1. Stopped ibus by running 'ibus exit' command. 2. Tried to install Keyman using km-config command. Verified that a message displays that "ibus-daemon is not running.." on the screen. 3. Pressed No button and verified that the Configuration windows exits. 4. Click the km-config again. Pressed Yes button from the warning message. 5. Installed a Keyman keyboard. Verified that we can switch to the keyboard and the keyboard works fine. 

   

* **TEST_NOIBUS_INST_CLI (PASSED):** 1. Stopped ibus by running 'ibus exit'. 2. Tried to install Khmer_Angkor keyboard by using the mentioned command. Verified that an error message is displaying on the terminal window. 3. Started ibus using  "ibus start" command. 4. Run the km-package-install again. Verified that the Khmer Angkor keyboard was successfully installed and we can switch to the keyboard and the keyboard works fine.

**Comment by bharanidharanj** [comment]
# Test Results

### GROUP_NOBLE_WAYLAND: Ubuntu 24.04 Noble with Gnome Shell and Wayland

* **TEST_REG_INST_UI (PASSED):** Tested with the attached PR build (17.0.279-beta (package version 17.0.279-1~PR-10788-2472.1+noble1) on Ubuntu 24.04 Noble OS (Wayland) (Virtual Box) and here is my observation: 1. Installed a Keyman keyboard using km-config command. 2. Verified that there are no messages about non-running ibus-daemon shows up. 3. Verified that it is possible to switch keyboard  and the keyboard is working fine. 

* **TEST_REG_INST_CLI (PASSED):** 1. Opened the terminal window. 2. Installed Khmer Angkor keyboard using the km-package-install command on the terminal. Verified that no message about non-running ibus-daemon shows up on ther terminal window. 3. Also, verified that we can switch to the keyboard and the keyboard is working fine. 

* **TEST_NOIBUS_INST_UI (PASSED):** 1. Stopped ibus by running 'ibus exit' command. 2. Tried to install Keyman using km-config command. Verified that a message displays that "ibus-daemon is not running.." on the screen. 3. Pressed No button and verified that the Configuration windows exits. 4. Click the km-config again. Pressed Yes button from the warning message. 5. Installed a Keyman keyboard. Verified that we can switch to the keyboard and the keyboard works fine. 

* **TEST_NOIBUS_INST_CLI (PASSED):** 1. Stopped ibus by running 'ibus exit'. 2. Tried to install Khmer_Angkor keyboard by using the mentioned command. Verified that an error message is displaying on the terminal window. 3. Started ibus using  "ibus start" command. 4. Run the km-package-install again. Verified that the Khmer Angkor keyboard was successfully installed and we can switch to the keyboard and the keyboard works fine.

**Comment by ermshiperete** [comment]
@bharanidharanj  Make sure you have the patched ibus version installed, then `ibus start` should work.

@keymanapp-test-bot retest GROUP_FOCAL TEST_NOIBUS_INST_CLI

**Comment by bharanidharanj** [comment]
# Test Results

### GROUP_FOCAL: Ubuntu 20.04 Focal with Gnome Shell and X11

* **TEST_NOIBUS_INST_CLI (PASSED):** notes

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.287-beta](https://keyman.com/downloads/releases/beta/17.0.287)

**Pull request merged**